src/test/java/lambda/part3/exercise/ReachIterableTest.java

```diff
@@ -46,32 +46,41 @@
             );
         }
 
-        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> function) {
-            final List<R> results = new ArrayList<>();
-            while (this.forNext(element -> results.addAll(function.apply(element))));
-            return ReachIterable.from(results);
+        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> mapper) {
+            final AtomicReference<ReachIterable<R>> inner = new AtomicReference<>(
+                    ReachIterable.from(Collections.emptyList())
+            );
+            final AtomicBoolean innerHasNext = new AtomicBoolean(false);
+            final Consumer<T> mapAndExtract = element -> inner.set(ReachIterable.from(mapper.apply(element)));
+            return consumer -> {
+                boolean outerHasNext = true;
+                if (!innerHasNext.get()) {
+                    outerHasNext = this.forNext(mapAndExtract);
+                }
+                innerHasNext.set(inner.get().forNext(consumer));
+                return  innerHasNext.get() || outerHasNext;
```

Comment:
Incorrect. You can return `true` without calling `consumer`. (`innerHasNext.get() == false && outerHasNext == true`)